File: cpp/src/VideoEncoder.cpp
#include <stdio.h>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <chrono>
#include <opencv2/opencv.hpp>
#include "Frame.cpp"
#include "BitStream.cpp"
#include "Golomb.cpp"
#include "Predictor.cpp"
#include "Encoders.cpp"

int main() {
    IntraFrameEncoder ife;
    Frame444* frame = new Frame444(720, 1280, "../../python/media/park_joy_444_720p50.y4m");
    JPEG1* predictor = new JPEG1();
    BitStream* bitstream = new BitStream("../out/park_joy_444_720p50.bin", 'w');
    Golomb* golomb = new Golomb(4);

    int total_time = 0;
    int total_frames = 0;
    bool firstFrame = true;

    while(true)
    {
        std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
        bool playing = frame->advance();

        // movie end
        if(!playing)
            break;
        
        vector<int> Y = frame->getY();
        ife.setMatrix(Y);
        if(firstFrame)
        {
            bitstream->writeString("500\t720\t1280");
            firstFrame = false;
        }
            
        ife.encode(frame, predictor, bitstream, golomb);

        vector<int> U = frame->getU();
        ife.setMatrix(U);
        ife.encode(frame, predictor, bitstream, golomb);

        vector<int> V = frame->getV();
        ife.setMatrix(V);
        ife.encode(frame, predictor, bitstream, golomb);
        
        total_frames++;
        std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
        std::cout << "Frame compressed in " << std::chrono::duration_cast<chrono::seconds>(end - begin).count() << " s;\n";
        
        // Com este break, só um frame é codificado
        // break;
    }

    return 0;
}

File: python/src/video-encoder.py
import datetime
import logging
import predictors
from frames import *
from encoders import IntraFrameEncoder

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger('root')
logger.setLevel(logging.INFO)

c_handler = logging.StreamHandler()
c_format = logging.Formatter('%(name)s - %(levelname)s - %(message)s')
c_handler.setFormatter(logging.Formatter('%(name)s | %(levelname)s ->: %(message)s'))

logger.addHandler(c_handler)
logger.propagate = False  # https://stackoverflow.com/a/19561320

"""
This program encodes 500 frames from the video "park_joy_444_720p50.y4m".
"""
if __name__ == '__main__':

    frame = Frame444(720, 1280, "../media/park_joy_444_720p50.y4m")

    total = 0
    firstFrame = True
    ife = IntraFrameEncoder(predictors.JPEG1)
    frames_no = 0
    while True:
        start = datetime.datetime.now()
        playing = frame.advance()

        # movie end
        if not playing:
            break

        # encode Y matrix
        matrix = frame.getY()
        ife.setMatrix(matrix)
        if firstFrame:
            firstFrame = False
            ife.bitstream.writeString("F500\tH720\tW1280\tC444\tS50")  # has to be concordant with Frame initialization above
        ife.encode()

        # encode U matrix
        matrix = frame.getU()
        ife.setMatrix(matrix)
        ife.encode()

        # encode V matrix
        matrix = frame.getV()
        ife.setMatrix(matrix)
        ife.encode()

        end = datetime.datetime.now() - start
        total += end.seconds
        frames_no += 1
        print("Frame compressed in {} s. Total bits: {}. Frames no. {}".format(end.seconds, ife.written_bits, frames_no))
        
        # com este break só codifica um frame
        if frames_no == 3:
        	break
        
    ife.bitstream.closeFile()
    print("Compressed frames in {} s.".format(total))